include/util/coordinate_calculation.hpp

```diff
@@ -25,16 +25,49 @@ const constexpr long double EARTH_RADIUS = 6372797.560856;
 
 
 //! Takes the squared euclidean distance of the input coordinates. Does not return meters!
-std::uint64_t squaredEuclideanDistance(const Coordinate &lhs, const Coordinate &rhs);
+std::uint64_t squaredEuclideanDistance(const Coordinate lhs, const Coordinate rhs);
 
 double haversineDistance(const Coordinate first_coordinate, const Coordinate second_coordinate);
 
 double greatCircleDistance(const Coordinate first_coordinate, const Coordinate second_coordinate);
 
-std::pair<double, FloatCoordinate>
-projectPointOnSegment(const FloatCoordinate &projected_xy_source,
-                      const FloatCoordinate &projected_xy_target,
-                      const FloatCoordinate &projected_xy_coordinate);
+inline std::pair<double, FloatCoordinate> projectPointOnSegment(const FloatCoordinate &source,
+                                                         const FloatCoordinate &target,
+                                                         const FloatCoordinate &coordinate)
+{
+    const FloatCoordinate slope_vector{target.lon - source.lon, target.lat - source.lat};
+    const FloatCoordinate rel_coordinate{coordinate.lon - source.lon, coordinate.lat - source.lat};
+    // dot product of two un-normed vectors
+    const auto unnormed_ratio = static_cast<double>(slope_vector.lon * rel_coordinate.lon) +
+                                static_cast<double>(slope_vector.lat * rel_coordinate.lat);
+    // squared length of the slope vector
+    const auto squared_length = static_cast<double>(slope_vector.lon * slope_vector.lon) +
+                                static_cast<double>(slope_vector.lat * slope_vector.lat);
+
+    if (squared_length < std::numeric_limits<double>::epsilon())
+    {
+        return {0, source};
+    }
+
+    const double normed_ratio = unnormed_ratio / squared_length;
+    double clamped_ratio = normed_ratio;
+    if (clamped_ratio > 1.)
+    {
+        clamped_ratio = 1.;
+    }
+    else if (clamped_ratio < 0.)
+    {
+        clamped_ratio = 0.;
+    }
```

Comment:
I searched the codebase for a `clamp` function and found one in `json_util.hpp` (which is never used by the way :P) --- worth it to provide a utility function for it?